### PR TITLE
Add option to disable users

### PR DESCRIPTION
**Pull request by jcpunk**
Rather than just setting the password to a random value, this extends the user management code to let a user be fully disabled.

### CODE REVIEW

**Comment by jcpunk** [comment]
When I run the tests locally they seem to pass...

**Comment by jcpunk** [comment]
I do not understand the test failures...

**Comment by jhoblitt** [comment]
@jcpunk I am able to reproduce the errors.

File: manifests/network.pp [hunk 1]
```diff
@@ -18,7 +18,7 @@
   Enum['dhcp', 'static'] $type = 'dhcp',
   Integer $lan_channel         = 1,
 ) {
-  require ipmi
+  require ipmi::ipmi::config
```
Comment (jhoblitt) [inline manifests/network.pp:21]: This appears to be a typo.

File: manifests/snmp.pp [hunk 2]
```diff
@@ -10,7 +10,7 @@
   String $snmp         = 'public',
   Integer $lan_channel = 1,
 ) {
-  require ipmi
+  require ipmi::config
```
Comment (jhoblitt) [inline manifests/snmp.pp:13]: `ipmi::config` needs the base class to be evaluated to provide `$ipmi::watchdog`. What are you trying to accomplish by not requiring the base class?

**Comment by jcpunk** [comment]
I think I've got the tests passing.  You can check the dep loop by pulling down this branch and dropping the `require ipmi::install` for `require ipmi` and running the test suite.

I'm showing:
```
dependency cycles found: (Exec[ipmi_user_add_newuser] => Ipmi::User[newuser] => Class[Ipmi] => Ipmi::User[newuser] => Exec[ipmi_user_add_newuser])
```

File: manifests/install.pp [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #
 # @api private
 #
-class ipmi::install {
+class ipmi::install inherits ipmi {
```
Comment (jhoblitt) [inline manifests/install.pp:4]: Inheriting the base class shouldn't be necessary.

File: manifests/user.pp [hunk 4]
```diff
@@ -13,63 +13,94 @@
...
+#   On SuperMicro IPMI, user id 2 is reserved for the 'ADMIN' username.
+#   On ASUS IPMI, user id 2 is reserved for the 'admin' username.
 #
 define ipmi::user (
-  Variant[Sensitive[String[1]], String[1]] $password,
   String $user     = 'root',
   Integer $priv    = 4,
+  Boolean $enable  = true,
   Integer $user_id = 3,
+  Optional[Variant[Sensitive[String[1]], String[1]]] $password = undef,
```
Comment (jhoblitt) [inline manifests/user.pp:28]: If we are going to change the order of the parameters, could we change the order of the puppet doc comments to be the same?

File: manifests/user.pp [hunk 5]
```diff
@@ -13,63 +13,94 @@
...
 
-  case $priv {
-    1: { $privilege = 'CALLBACK' }
-    2: { $privilege = 'USER' }
-    3: { $privilege = 'OPERATOR' }
-    4: { $privilege = 'ADMINISTRATOR' }
-    default: { fail('invalid privilege level specified') }
-  }
+  if $enable {
+    if empty($password) {
```
Comment (jhoblitt) [inline manifests/user.pp:33]: Is this check necessary? I believe the type annotation should prevent `$password` from being an empty string.

File: manifests/user.pp [hunk 6]
```diff
@@ -13,63 +13,94 @@
...
-  }
+    exec { "ipmi_user_setpw_${title}":
+      command => "/usr/bin/ipmitool user set password ${user_id} \'${real_password}\'",
+      unless  => "/usr/bin/ipmitool user test ${user_id} 16 \'${real_password}\'",
+      notify  => [Exec["ipmi_user_enable_${title}"], Exec["ipmi_user_enable_sol_${title}"], Exec["ipmi_user_channel_setaccess_${title}"]],
+    }
+
+    exec { "ipmi_user_enable_sol_${title}":
+      command     => "/usr/bin/ipmitool sol payload enable 1 ${user_id}",
+      refreshonly => true,
```
Comment (jhoblitt) [inline manifests/user.pp:77]: Do the refershonly exec resources need to be inside of the `$enable` if?  There is a lot going on here and I'm wondering if cutting down the size of the if blocks whould make it easier to understand the logic.

**Review by jhoblitt — CHANGES_REQUESTED** [review]

File: manifests/install.pp (hunk 3 above)
Comment (jcpunk) [reply manifests/install.pp:4]: Without this inheritance, the unit tests for the `define` fail as the toplevel class vars aren't instantiated.

File: manifests/user.pp (hunk 4 above)
Comment (jcpunk) [reply manifests/user.pp:28]: Done.  I re-arranged these to make puppet-lint happy.

File: manifests/user.pp (hunk 5 above)
Comment (jcpunk) [reply manifests/user.pp:33]: With `enable==false` the `password` isn't required, the type annotation changed to permit omitting the password.  But there doesn't seem to be a way in the type defs to say "if X==true then Y must be Z"

File: manifests/user.pp (hunk 6 above)
Comment (jcpunk) [reply manifests/user.pp:77]: I was thinking keeping them within the `if` scope helped make it clear these elements are part of the same logical block.  I'm open to moving them out, but having them under the same `{}` seemed to be a clean blob.

File: manifests/install.pp (hunk 3 above)
Comment (jhoblitt) [reply manifests/install.pp:4]: This works for me: 

```puppet
diff --git a/manifests/install.pp b/manifests/install.pp
index 88db7fa..5980fec 100644
--- a/manifests/install.pp
+++ b/manifests/install.pp
@@ -1,8 +1,10 @@
 #
 # @api private
 #
-class ipmi::install inherits ipmi {
+class ipmi::install {
   assert_private()
 
+  include ipmi
+
   ensure_packages($ipmi::packages)
 }
```

Comment (jcpunk) [reply manifests/install.pp:4]: When I do that, I get a dep loop on the top level create resources for `ipmi::users`.

Comment (jhoblitt) [reply manifests/install.pp:4]: Did you use `include` or `require` for you test?

Comment (jcpunk) [reply manifests/install.pp:4]: Add, I did `require` that would probably do it.

Comment (jcpunk) [reply manifests/install.pp]: Fixed.

**Comment by jcpunk** [comment]
Just checking back in.

**Comment by jhoblitt** [comment]
Sorry for the slow response. I wanted to try to test this a box with a bmc before merging it.  It looks like the gha failure on centos 8 is bogus. Could you touch the branch to retrigger it? It isn't giving me the re-run button.

**Comment by jcpunk** [comment]
poked and re-pushed

**Comment by jcpunk** [comment]
Any guesses why it doesn't like `REFERENCE.md`? I'm showing it is the current copy.

**Comment by jhoblitt** [comment]
@jcpunk running `bundle exec rake reference` on this branch does result in changes for me.

**Comment by jcpunk** [comment]
I've been running `puppet strings generate --format markdown --out REFERENCE.md` is that not the way?

**Review by jhoblitt — APPROVED** [review]
LGTM

**Pull request merged**